### PR TITLE
feat: dashboard renderer grid layout

### PR DESCRIPTION
**Pull request by filipgutica**
# Summary

Grid layout
https://konghq.atlassian.net/browse/MA-2268

Updated `DashboardRenderer` to use `GridLayout` for laying out the `Tiles` 

Updated `DashboardDefinition` and `TileDefinition` schemas to include positioning and sizing info that is passed off to the `GridLayout` component for display.

See the following example:
```vue
<DashboardRenderer
  :context="context"
  :config="config"
/>
```
```typescript
const context: DashboardRendererContext = {
  filters: [],
  timeSpec: {
    type: 'relative',
    time_range: '15M',
  },
}

const config: DashboardConfig = {
  // 5 x 5 grid
  gridSize: {
    cols: 5,
    rows: 5,
  }
  tiles: [
    {
      definition: {
        chart: {
          type: ChartTypes.HorizontalBar,
        },
        query: {},
      },
      layout: {
        // Position at column 0, row 0
        position: {
          col: 1,
          row: 1,
        },
        // Spans 3 columns and 2 rows
        size: {
          col: 3,
          row: 2,
        }
      }
    },
    {
      definition: {
        chart: {
          type: ChartTypes.Gauge,
          metricDisplay: ChartMetricDisplay.Full,
          reverseDataset: true,
          numerator: 0,
        },
        query: {},
      },
      layout: {
        // Position at column 0, row 2
        position: {
          col: 0,
          row: 2,
        },
        // Spans 3 columns and 2 rows
        size: {
          col: 3,
          row: 2,
        }
      }
    },
  ],
}
```

## PR Checklist

* [ ] **Naming & Structure:** the files and package structure use the conventions outlined in the [Creating a Package docs](https://github.com/Kong/public-ui-components/blob/main/docs/creating-a-package.md).
* [ ] **Tests pass:** check the output of all package unit and/or component tests.
  * [ ] If this PR is the result of a bug, test coverage was added accordingly.
* [ ] **Functional:** all changes do not break existing APIs, but if so, a BREAKING CHANGE commit is in place to bump the major version.
* [ ] **Conventional Commits** all commits follow the conventional commit standards [outlined in the main README](https://github.com/Kong/public-ui-components#committing-changes).
* [ ] **Docs:** includes a technically accurate README, and the docs have been updated accordingly based on the changes in this PR.

### CODE REVIEW

File: packages/analytics/dashboard-renderer/src/types/dashboard-renderer-types.ts [hunk 1]
```diff
@@ -28,6 +30,18 @@ export const barChartSchema = {
     stacked: {
       type: 'boolean',
     },
+    showAnnotations: {
+      type: 'boolean',
+    },
+    chartDatasetColors: {
```
Comment (adorack) [inline packages/analytics/dashboard-renderer/src/types/dashboard-renderer-types.ts:36]: Could you please refactor this definition out into its own constant?  It's shared between bar charts and timeseries charts.

**Review by adorack — CHANGES_REQUESTED** [review]
Only one minor change request.

**Comment by adorack** [comment]
Also, minor: typo in commit: "layout" instead of "playout".

File: packages/analytics/dashboard-renderer/src/components/AnalyticsChartRenderer.vue [hunk 2]
```diff
@@ -1,32 +1,39 @@
...
 const chartTypeLookup = {
   [ChartTypes.HorizontalBar]: AnalyticsChartTypes.HORIZONTAL_BAR,
   [ChartTypes.VerticalBar]: AnalyticsChartTypes.VERTICAL_BAR,
+  [ChartTypes.TimeseriesLine]: AnalyticsChartTypes.TIMESERIES_LINE,
 }
 
-const options = computed(() => ({
+const options = computed<AnalyticsChartOptions>(() => ({
   type: chartTypeLookup[props.chartOptions.type],
+  fill: (props.chartOptions as TimeseriesChartOptions).fill,
```
[the remark this answers is not in the document]
Comment (filipgutica) [reply packages/analytics/dashboard-renderer/src/components/AnalyticsChartRenderer.vue:36]: May be we can use generics here too. 

It's not really an issue, seems to handle the property being there vs not being there gracefully. I'm not a huge fan of the casting though either.

[the remark this answers is not in the document]
Comment (filipgutica) [reply packages/analytics/dashboard-renderer/src/components/AnalyticsChartRenderer.vue]: I ended up splitting it into separate renderer components to keep things "modular"

File: packages/analytics/dashboard-renderer/src/components/layout/GridLayout.vue [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    >
+      <slot
+        name="tile"
+        :tile="cell.tile"
+      />
+    </div>
+  </div>
+</template>
+
+<script lang="ts" setup generic="T">
```
[the remark this answers is not in the document]
Comment (filipgutica) [reply packages/analytics/dashboard-renderer/src/components/layout/GridLayout.vue:23]: Yes, I saw your post about that in the Vue channel, I like this feature a lot 😄

File: packages/analytics/dashboard-renderer/src/components/BarChartRenderer.vue [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
       chart-title=""
+      :height="`${height}px`"
+      legend-position="bottom"
+      :show-annotations="chartOptions.showAnnotations"
       tooltip-title=""
     />
   </QueryDataProvider>
 </template>
-<script setup lang="ts">
+<script setup lang="ts" generic="">
```
Comment (adorack) [inline packages/analytics/dashboard-renderer/src/components/BarChartRenderer.vue:18]: Minor: remove `generic` attribute for the bar chart renderer and timeseries renderer.

File: packages/analytics/dashboard-renderer/src/components/AnalyticsChartRenderer.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (adorack) [reply packages/analytics/dashboard-renderer/src/components/AnalyticsChartRenderer.vue]: Ah, that looks good -- I thought there'd be a lot of code duplication, but there actually isn't much.

**Review by adorack — DISMISSED** [review]

File: packages/analytics/dashboard-renderer/src/mock-data.ts [hunk 5]
```diff
@@ -33,3 +33,252 @@ export const exploreV3Response = {
     timestamp: '2023-10-15T05:00:00.000Z',
   }, { event: { STATUS_CODE_GROUPED: '5XX', REQUEST_COUNT: 17275256 }, timestamp: '2023-10-15T05:00:00.000Z' }],
 }
+
+export const timeSeriesExploreResponse = {
```
Comment (adamdehaven) [inline packages/analytics/dashboard-renderer/src/mock-data.ts:37]: issue: mock data should not go inside the source folder; this file will be bundled

**Comment by ValeryG** [comment]
Just out of curiosity, why isn't css grid (or even flex) layout is not enough? what justifies to have something home made here?

**Comment by filipgutica** [comment]
> > > Just out of curiosity, why isn't css grid (or even flex) layout is not enough? what justifies to have something home made here?
> > 
> > 
> > It's just not flexible enough for our needs.
> 
> Is there any document that summarizes our needs and justifies that whatever our needs are cannot be accomplished without home brewing grid implementation?

No but we've discussed, this, and re-discussed this several times in the past couple weeks. 

I'll summarize in a few key points

- Our requirements are a component that accepts a JSON blob representing a dashboard, what is on that dashboard, where those things are in the dashboard etc.
- Using css flex/grid is not flexible enough in creating a dashboard/grid powered by positional information like we want to do. On top of that there are performance issues (need to use nested loops etc especially if trying to use CSS grid). Further, the most widely used grid layout libraries use a similar approach of absolutely positioned cells/tiles, probably due to the same issues I ran into.
- There is a lack of good third party libraries that achieve this in Vue. The only viable library I found is a React one. The Vue libraries are buggy, poorly maintained and poorly supported.
- The decision was made to use a home grown solution for now as it more than satisfies our current requirements for dashboards until more time and resources can be poured into finding the most "correct" solution.

**Comment by filipgutica** [comment]
> > > Just out of curiosity, why isn't css grid (or even flex) layout is not enough? what justifies to have something home made here?
> > 
> > 
> > It's just not flexible enough for our needs.
> 
> Is there any document that summarizes our needs and justifies that whatever our needs are cannot be accomplished without home brewing grid implementation?

Also,  we don't intend to long term use a home brewed grid system. As we've discussed with @adamdehaven  and @jillztom, we'd like core ui to shepherd in a grid system for Konnect more broadly. When that's available and mature enough, we will switch over to whatever the recommended pattern is.
 For now however, given the poor other options we've found, we think this gives us the best path to hit our other objectives while that develops.

**Review by mihai-peteu — APPROVED** [review]

**Review by jillztom — DISMISSED** [review]

**Review by jillztom — APPROVED** [review]

**Review by mihai-peteu — APPROVED** [review]

**Pull request merged**